dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp

```diff
@@ -53,69 +56,95 @@ void ExecuteScalarSubqueriesMatcher::visit(ASTPtr & ast, Data & data)
+                ast_new->setAlias(ast->tryGetAlias());
+                ast = std::move(ast_new);
+                return;
+            }
+
+            if (block.rows() != 1 || res.in->read())
+                throw Exception("Scalar subquery returned more than one row", ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY);
+        }
+        catch (const Exception & e)
+        {
+            if (e.code() == ErrorCodes::TOO_MANY_ROWS)
+                throw Exception("Scalar subquery returned more than one row", ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY);
+            else
+                throw;
         }
 
-        if (block.rows() != 1 || res.in->read())
-            throw Exception("Scalar subquery returned more than one row", ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY);
-    }
-    catch (const Exception & e)
-    {
-        if (e.code() == ErrorCodes::TOO_MANY_ROWS)
-            throw Exception("Scalar subquery returned more than one row", ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY);
+        block = materializeBlock(block);
+        size_t columns = block.columns();
+
+        if (columns == 1)
+            scalar = block;
         else
-            throw;
+        {
+
+            ColumnWithTypeAndName ctn;
+            ctn.type = std::make_shared<DataTypeTuple>(block.getDataTypes());
+            ctn.column = ColumnTuple::create(block.getColumns());
+            scalar.insert(ctn);
+        }
     }
 
-    size_t columns = block.columns();
-    if (columns == 1)
+    if (worthConvertingToLiteral(scalar))
     {
-        auto lit = std::make_unique<ASTLiteral>((*block.safeGetByPosition(0).column)[0]);
+        auto lit = std::make_unique<ASTLiteral>((*scalar.safeGetByPosition(0).column)[0]);
         lit->alias = subquery.alias;
         lit->prefer_alias_to_column_name = subquery.prefer_alias_to_column_name;
-        ast = addTypeConversionToAST(std::move(lit), block.safeGetByPosition(0).type->getName());
+        ast = addTypeConversionToAST(std::move(lit), scalar.safeGetByPosition(0).type->getName());
     }
     else
     {
-        auto tuple = std::make_shared<ASTFunction>();
-        tuple->alias = subquery.alias;
-        ast = tuple;
-        tuple->name = "tuple";
-        auto exp_list = std::make_shared<ASTExpressionList>();
-        tuple->arguments = exp_list;
-        tuple->children.push_back(tuple->arguments);
-
-        exp_list->children.resize(columns);
-        for (size_t i = 0; i < columns; ++i)
-        {
-            exp_list->children[i] = addTypeConversionToAST(
-                std::make_unique<ASTLiteral>((*block.safeGetByPosition(i).column)[0]),
-                block.safeGetByPosition(i).type->getName());
-        }
+        auto func = makeASTFunction("__getScalar", std::make_shared<ASTLiteral>(scalar_query_hash_str));
+        func->alias = subquery.alias;
+        func->prefer_alias_to_column_name = subquery.prefer_alias_to_column_name;
```

Comment:
```prefer_alias_to_column_name``` is a hack and deprecated for code that can be done without it. Can't we avoid it?

Reply:
hmm, I do see some references to `prefer_alias_to_column_name`. I'm not familiar with that. Perhaps it deserves a separate pr to clean them all.